### DISCUSSION TITLE
Batch recover files in the recycle bin

### DISCUSSION THREAD

**Discussion by T-TRz879** [juicedata/juicefs#2412] [Ideas]
I think this function is very helpful for misoperation `rm -rf *`,But facing many problems, I sorted out several problems as follows:
1. The recovered parent directory does not exist
2. The revovered file/dir does exist
3.  The parent directory of the recovered file is also in the recycle bin, but not in the same time subTrash folder.When the file and parent directory are in the same subTrash, it can recover the folder first
4. When recovered a file, cleanupTrash thread is working.it must conflict and will lead to incomplete recovery

**Comment by davies** [comment]
Yes, we could have a tool to recover a file or a directory recursively.

**Reply by T-TRz879** [reply]
If I can, I want to participate in the implementation of this function!!

**Reply by T-TRz879** [reply]
@xiaogaozi

**Reply by xiaogaozi** [reply]
You are very welcome to contribute code to the JuiceFS community, please read the [Contributing Guide](https://juicefs.com/docs/community/development/contributing_guide) document for related information.

**Reply by T-TRz879** [reply]
ok,Thanks! :)

**Comment by T-TRz879** [comment]
@davies @xiaogaozi @Suave @suzaku 
I wrote the design idea of batch recover. You can give me some advice.
[批量恢复.md](https://github.com/juicedata/juicefs/files/9270317/default.md)

**Comment by qingyunha** [comment]
Could we make the trash name exactly as it is, with the path included, so that we can just use "mv" to recover?

**Reply by T-TRz879** [reply]
I understand what you mean. The advantage of your idea is that you can let the user see the location before this file. I think that in this case, the value of the **sub file / directory** saved by `subtrash` should be saved in `inode` and `attr`. This can facilitate our `mv`. When we `unlink`, `rename` and `rmdir`, we need to consider the duplicate file names in `subtrash`
